Add user guide to new Fast Pair enable API (#14170)
nrfconnect/sdk-nrf: doc/nrf/external_comp/bt_fast_pair.rst

You must enable the :kconfig:option:`CONFIG_BT_FAST_PAIR` Kconfig option to support the Google Fast Pair standard in your project.

> The Fast Pair subsystem requires enabling Bluetooth with the :c:func:`bt_enable` function and loading Zephyr's :ref:`zephyr:settings_api` with the :c:func:`settings_load` function.

Reviewer: Maybe we can put this section in the proximity of L171, as you don't need to perform these operations before calling register APIs. You just need to prepare these dependencies before you enable the Fast Pair subsystem:

```
After the callback registration, the Fast Pair subsystem must be enabled with the :c:func:`bt_fast_pair_enable` function.
Before the :c:func:`bt_fast_pair_enable` operation, you must enable Bluetooth with the :c:func:`bt_enable` function and load Zephyr's :ref:`zephyr:settings_api` with the :c:func:`settings_load` function.
```

Set your application-specific callback functions in the callback structure that is the input parameter for the ``..._register`` API function.
The callback structure must persist in the application memory (static declaration), as during the registration, the Fast Pair module stores only the memory pointer to it.

The standard Fast Pair API (without extensions) currently supports the :c:func:`bt_fast_pair_info_cb_register` function (optional) for registering application callbacks.

The standard Fast Pair (without extensions) does not require registration of any callback type, meaning all callbacks are optional.

After the callback registration, the Fast Pair subsystem must be enabled with the :c:func:`bt_fast_pair_enable` function.
Before performing the :c:func:`bt_fast_pair_enable` operation, you must enable Bluetooth with the :c:func:`bt_enable` function and load Zephyr's :ref:`zephyr:settings_api` with the :c:func:`settings_load` function.
The Fast Pair subsystem readiness can be checked with the :c:func:`bt_fast_pair_is_ready` function.
The Fast Pair subsystem can be disabled with the :c:func:`bt_fast_pair_disable` function.

> In the Fast Pair subsystem disabled state, most of the Fast Pair APIs are not available.

Reviewer 2: Seems like the second part of @kapi-no's comment was overlooked. Consider adding it as follows:

`The Fast Pair GATT Service is statically defined, so it's still present in GATT database after Fast Pair is disabled. In the Fast Pair subsystem disabled state, GATT operations on the Fast Pair service are rejected.

The standard Fast Pair (without extensions) does not require registration of any callback type, meaning all callbacks are optional.

After the callback registration, the Fast Pair subsystem must be enabled with the :c:func:`bt_fast_pair_enable` function.
The Fast Pair subsystem readiness can be checked with the :c:func:`bt_fast_pair_is_ready` function.

> The Fast Pair subsystem can be disabled with the :c:func:`bt_fast_pair_disable` function.

Reviewer: Consider also mentioning the effects of having the subsystem in the disabled state:

- One sentence here: `In the Fast Pair subsystem disabled state, most Fast Pair APIs are not available`

- One sentence in the Interacting with GATT service section. For example: `In the Fast Pair subsystem disabled state, GATT operations on the Fast Pair service are rejected`.

> Apart from the callback registration and enabling the Fast Pair subsystem, no additional operations are needed to integrate the standard Fast Pair implementation.

Reviewer 2: Consider mentioning that user needs to enable Bluetooth and load settings before enabling Fast Pair.

Reviewer 2: `to integrate the core Fast Pair implementation`?

Reviewer: > `to integrate the core Fast Pair implementation`?

We have a few mentions (L165, L167) of the standard Fast Pair implementation in this RST file. Calling it here the core Fast Pair implementation would confuse the user. We agreed in previous code reviews to use the "standard" keyword to refer to the core part. I think we should keep it as it is

Reviewer: My comment was about keeping the consistency. The confusion here stems from the fact that we use two different words to refer to the same thing.

Is there a `core` use case? Both "core" and "standard" are made-up words by us and are not used in the original Fast Pair specification. When I say that we agreed to the previous wording, I have the following PR in mind:

https://github.com/nrfconnect/sdk-nrf/pull/13229

where there were no comments regarding this wording

I will leave it up to @alstrzebonski to select the approach for this one. However, if we decide to go with the "core" word, then we need to align it consistently in this guide.

Author: I will leave it as is. I think it's not confusing and reader can easily understand what we mean.

The procedure is initiated by the Fast Pair Seeker after Bluetooth LE connection is established.
No application interaction is required.

> In the Fast Pair subsystem disabled state, GATT operations on the Fast Pair service are rejected.

Reviewer 2: The last small improvement (would be good to explain why it's implemented like that I think)

```
The Fast Pair GATT Service is statically defined, so it's still present in GATT database after Fast Pair is disabled. In the Fast Pair subsystem disabled state, GATT operations on the Fast Pair service are rejected.
```

Author: Addressed.

The procedure is initiated by the Fast Pair Seeker after Bluetooth LE connection is established.
No application interaction is required.

> The Fast Pair GATT Service is statically defined, so it's still present in the GATT database after the Fast Pair subsystem is disabled.

Reviewer 3: ```suggestion
The Fast Pair GATT service is statically defined, so it is still present in the GATT database after the Fast Pair subsystem is disabled.
```

The Fast Pair GATT service uses a non-volatile memory to store the Fast Pair user data such as Account Keys and the Personalized Name.
This data can be cleared by calling the :c:func:`bt_fast_pair_factory_reset` function.

> Calling the :c:func:`bt_fast_pair_factory_reset` function doesn't affect the Fast Pair subsystem readiness.

Reviewer 3: ```suggestion
Calling the :c:func:`bt_fast_pair_factory_reset` function does not affect the Fast Pair subsystem's readiness.
```

> If the subsystem is enabled with the :c:func:`bt_fast_pair_enable` function, it will stay enabled after calling the :c:func:`bt_fast_pair_factory_reset` function.

Reviewer 3: ```suggestion
If the subsystem is enabled with the :c:func:`bt_fast_pair_enable` function, it stays enabled after calling the :c:func:`bt_fast_pair_factory_reset` function.
```

Ensure that your reset action implementation executes correctly in the following execution contexts:

> * In the :c:func:`bt_fast_pair_factory_reset` function context: the factory reset action is triggered by calling the :c:func:`bt_fast_pair_factory_reset` function.

Reviewer 3: ```suggestion
* In the :c:func:`bt_fast_pair_factory_reset` function context - The factory reset action is triggered by calling the :c:func:`bt_fast_pair_factory_reset` function.
```

> * In the :c:func:`bt_fast_pair_enable` function context: the factory reset action using the :c:func:`bt_fast_pair_factory_reset` function was interrupted and the factory reset is retried when enabling the Fast Pair subsystem.

Reviewer 3: ```suggestion
* In the :c:func:`bt_fast_pair_enable` function context - The factory reset action using the :c:func:`bt_fast_pair_factory_reset` function was interrupted, and the factory reset is retried when enabling the Fast Pair subsystem.
```
